BDArmory/UI/BDAEditorArmorWindow.cs

```diff
@@ -489,6 +513,35 @@ void WindowArmor(int windowID)
                     CalculateArmorMass();
                 }
                 line += 0.5f;
+                resourcePick = GUI.Toggle(new Rect(10, (line++ + armorLines) * lineHeight, 280, lineHeight), resourcePick, StringUtils.Localize("Drymass Resources"), resourcePick ? BDArmorySetup.BDGuiSkin.box : BDArmorySetup.BDGuiSkin.button);
+                if (resourcePick)
+                {
+                    int pos = 0;
+                    line++;
+                    using (var res = vesselResources.GetEnumerator())
+                        while (res.MoveNext())
+                        {
+                            if (res.Current.density == 0) continue; //don't show massless resouces for drymass blacklist
+                            if (res.Current.name.Contains("Intake")) continue; //don't include intake air, since that will always be present                            
+                            int resID = PartResourceLibrary.Instance.GetDefinition(res.Current.name).id;
+                            if (GUI.Button(new Rect(10 + (pos % 2) * (280) / 2f, (line + (int)(pos / 2)) * lineHeight, (280) / 2f, lineHeight), $"{res.Current.displayName}", vesselResourceIDs.Contains(resID) ? BDArmorySetup.BDGuiSkin.button :BDArmorySetup.BDGuiSkin.box))
```

Comment:
I find the default way this works to be a bit unintuitive.
I think it'd be better if the selected dry mass resources mean that those resources are considered when calculating the dry mass, so that dry mass is with those selected resources drained.

Also, I think the default should be to select all newly added resources. This would need `OnEditorShipModifiedEvent` to detect when new resources are added, which could be done by modifying it with
```diff
diff --git a/BDArmory/UI/BDAEditorArmorWindow.cs b/BDArmory/UI/BDAEditorArmorWindow.cs
index 88e29a5f3..ed3f28e05 100644
--- a/BDArmory/UI/BDAEditorArmorWindow.cs
+++ b/BDArmory/UI/BDAEditorArmorWindow.cs
@@ -159,2 +159,3 @@ private void OnEditorShipModifiedEvent(ShipConstruct data)
                 StartCoroutine(DelayedRefreshVisuals(data));
+            var oldResources = vesselResources.ToHashSet();
             vesselResources.Clear();
@@ -171,2 +172,11 @@ private void OnEditorShipModifiedEvent(ShipConstruct data)
                 }
+            var newResources = vesselResources.ToHashSet();
+            newResources.ExceptWith(oldResources); // Newly added resources.
+            var resourceIDs = newResources.Select(res => PartResourceLibrary.Instance.GetDefinition(res.name).id).ToHashSet();
+            resourceIDs.ExceptWith(vesselResourceIDs.ToHashSet()); // Only newly added resources that aren't already added to the IDs list.
+            if (resourceIDs.Count > 0)
+            {
+                vesselResourceIDs.AddRange(resourceIDs);
+                if (!FerramAerospace.hasFAR) CalculateTotalLift();
+            }
         }
```

Reply:
The request from FJRT was to have something that showed drymass with filtered resources included, similar to RCSBuildAid; e.g. a plane with 1000kg of LiquidFuel and 254kg of 20x102mm, you could have the drymass of the airframe include that 254kg of ammo, with only the fuel counting towards wet mass, without having to go though the hassle of manually draining a bunch of resource containers.
If selected resources need to be drained to count for drymass (so full tanks of selected resources are still counting for wet mass, then?), then what's the point of filtering them in the first place, that's no different from current behavior.

Reply:
Changes made; resources are all selected at start, and by default will be treated as normal, so default drymass is true drymass; selecting specific resources will add their mass to the craft drymass; button label has been updated to be less ambiguous.

On a related note, does ``EditorLogic.fetch.ship.GetConnectedResourceTotals(resID, sim, out double amount, out double maxAmount)`` return amount/maxAmount values on your end? It's returning 0 on mine, causing wet/dry mass to report the same.

Reply:
`EditorLogic.fetch.ship.GetConnectedResourceTotals` is working for me. Adding a debug button for this with
```csharp
if (HighLogic.LoadedSceneIsEditor && GUI.Button(SLineRect(++line), "Test GetConnectedResourceTotals"))
{
    float dMass = 0;
    foreach (var res in EditorLogic.fetch.ship.parts.SelectMany(p => p.Resources, (p, r) => r.info).ToHashSet()) // Unique resource infos on the ship.
    {
        EditorLogic.fetch.ship.GetConnectedResourceTotals(res.id, true, out double fuelCurrent, out double fuelMax);
        dMass -= (float)fuelCurrent * res.density;
        Debug.Log($"DEBUG res {res.name}, ID: {res.id}, current: {fuelCurrent}, max: {fuelMax}, mass: {(float)fuelCurrent * res.density}");
    }
    Debug.Log($"DEBUG dMass: {dMass}");
    dMass = 0;
    foreach (var part in EditorLogic.fetch.ship.parts)
        foreach (var res in part.Resources)
        {
            dMass -= (float)res.amount * res.info.density;
        }
    Debug.Log($"DEBUG dMass: {dMass}");
}
```
gave
```
[LOG 12:19:08.207] DEBUG res ElectricCharge, ID: 1566956177, current: 50, max: 50, mass: 0
[LOG 12:19:08.207] DEBUG res LiquidFuel, ID: -1483389306, current: 200.000009536743, max: 1240, mass: 1
[LOG 12:19:08.208] DEBUG res 20x102Ammo, ID: -1023478857, current: 2600, max: 2600, mass: 0.6734
[LOG 12:19:08.208] DEBUG res 25x137Ammo, ID: 2044613620, current: 1250, max: 1250, mass: 0.66
[LOG 12:19:08.209] DEBUG res IntakeAir, ID: 1669081118, current: 0, max: 15, mass: 0
[LOG 12:19:08.209] DEBUG res IntakeAtm, ID: 197105542, current: 0, max: 15, mass: 0
[LOG 12:19:08.210] DEBUG dMass: -2.3334
[LOG 12:19:08.210] DEBUG dMass: -2.3334
```
for one of my craft that has 100LF in each of two wings.

As a side note, there's no need for most of the `PartResourceLibrary.Instance.GetDefinition(res.info.name)` calls as they're just returning `res.info`.

Reply:
You can even condense the `dMass` calculation into a one-liner if the above works for you:
```csharp
var dMass = -EditorLogic.fetch.ship.parts.SelectMany(p => p.Resources, (p, r) => r.info).ToHashSet().Select(res => { EditorLogic.fetch.ship.GetConnectedResourceTotals(res.id, true, out double fuelCurrent, out double fuelMax); return (float)fuelCurrent * res.density; }).Sum();
```

Reply:
The button spacing still looks a bit off. I think there's probably some missing/extra `line += 0.5f;` in there somewhere.

Reply:
These tweaks to the layout give a more consistent spacing and adds a shadow box behind the drymass toggles.
```diff
diff --git a/BDArmory/UI/BDAEditorArmorWindow.cs b/BDArmory/UI/BDAEditorArmorWindow.cs
index 534c087c3..a3f559aa7 100644
--- a/BDArmory/UI/BDAEditorArmorWindow.cs
+++ b/BDArmory/UI/BDAEditorArmorWindow.cs
@@ -449,3 +449,2 @@ void WindowArmor(int windowID)
                 }
-                line += 0.75f;
                 GUI.Label(new Rect(10, line * lineHeight, 300, lineHeight), $"{StringUtils.Localize("#LOC_BDArmory_ArmorTotalMass")}: {totalArmorMass:0.00}", style);
@@ -468,2 +467,3 @@ void WindowArmor(int windowID)
 #if DEBUG
+                line += 0.5f;
                 if (GUI.Button(new Rect(10, line++ * lineHeight, 280, lineHeight), "Find Wings", BDArmorySetup.ButtonStyle))
@@ -487,2 +487,3 @@ void WindowArmor(int windowID)
             {
+                line += 0.5f;
                 resourcePick = GUI.Toggle(new Rect(10, line++ * lineHeight, 280, lineHeight), resourcePick, StringUtils.Localize("#LOC_BDArmory_DryMassWhitelist"), resourcePick ? BDArmorySetup.BDGuiSkin.box : BDArmorySetup.BDGuiSkin.button);
@@ -490,2 +491,3 @@ void WindowArmor(int windowID)
                 {
+                    GUI.Box(new Rect(10, line * lineHeight - 2, 280, Mathf.CeilToInt(vesselResourceIDs.Count / 2f - 1) * lineHeight + 6), "", BDArmorySetup.BDGuiSkin.box); // l,r,t,b = 3,3,3,3 with slight overlap of the toggle
                     int pos = 0;
@@ -497,3 +499,3 @@ void WindowArmor(int windowID)
                             int resID = PartResourceLibrary.Instance.GetDefinition(res.Current.name).id;
-                            if (GUI.Button(new Rect(10 + (pos % 2) * (280) / 2f, (line + (int)(pos / 2)) * lineHeight, (280) / 2f, lineHeight), $"{res.Current.displayName}", vesselResourceIDs.Contains(resID) ? BDArmorySetup.BDGuiSkin.button : BDArmorySetup.BDGuiSkin.box))
+                            if (GUI.Button(new Rect(pos % 2 == 0 ? 13 : 152f, (line + (int)(pos / 2)) * lineHeight + 1, 135, lineHeight), $"{res.Current.displayName}", vesselResourceIDs.Contains(resID) ? BDArmorySetup.BDGuiSkin.button : BDArmorySetup.BDGuiSkin.box)) // match BDGUIComboBox's layout
                             {
@@ -511,5 +513,4 @@ void WindowArmor(int windowID)
                         }
-                    line += Mathf.CeilToInt(pos / 2f);
+                    line += Mathf.CeilToInt(pos / 2f) + 0.25f;
                 }
-                line += 0.5f;
             }
@@ -554,3 +555,2 @@ void WindowArmor(int windowID)
                 }
-                line += 0.5f;
 
@@ -558,2 +558,3 @@ void WindowArmor(int windowID)
                 {
+                    line += 0.5f;
                     ArmorStats = GUI.Toggle(new Rect(10, (line + armorLines) * lineHeight, 280, lineHeight), ArmorStats, StringUtils.Localize("#LOC_BDArmory_ArmorStats"), ArmorStats ? BDArmorySetup.BDGuiSkin.box : BDArmorySetup.BDGuiSkin.button);
```
![Screenshot_20240820_143329](https://github.com/user-attachments/assets/f14a938c-1c20-4861-818a-d204d0fd30ae)

The radar absorbent armour/hull don't fit their boxes though.